### PR TITLE
Fix pointwise loss broadcast

### LINKED ISSUE #12129: Broadcasting fails for mse_loss when tensors have no grad
## To Reproduce

```python
a = T.zeros(5,1,10, requires_grad=True)
b = T.zeros(5,10,10, requires_grad=True)
F.mse_loss(a,b)
# OK

a = T.zeros(5,1,10)
b = T.zeros(5,10,10)
F.mse_loss(a,b)
# RuntimeError: input and target shapes do not match: input [5 x 1 x 10], target [5 x 10 x 10] at /opt/conda/conda-bld/pytorch_1524577523076/work/aten/src/THNN/generic/MSECriterion.c:13
```

## Expected behavior

Broadcast should work the same as with grad.

## Workaround

```python
# assumes size_average=False
def my_mse(a,b):
    return ((a-b)**2).sum()
```

### PR DESCRIPTION
**Pull request by ailzhang**
Fixes #12129 , #12327 
cc: @fmassa `_pointwise_loss` is currently only called by `l1_loss`, `mse_loss` and `smooth_l1_loss` which all makes sense to have broadcast.

### CODE REVIEW

File: torch/nn/functional.py [hunk 1]
```diff
@@ -1785,10 +1786,10 @@ def smooth_l1_loss(input, target, size_average=None, reduce=None, reduction='ele
     See :class:`~torch.nn.SmoothL1Loss` for details.
     """
     if size_average is not None or reduce is not None:
-        reduction = _Reduction.legacy_get_enum(size_average, reduce)
-    else:
-        reduction = _Reduction.get_enum(reduction)
-    return torch._C._nn.smooth_l1_loss(input, target, reduction)
+        reduction = _Reduction.legacy_get_string(size_average, reduce)
+    return _pointwise_loss(
```
Comment (colesbury) [inline torch/nn/functional.py]: This looks like it's going to make smooth_l1_loss much slower. It goes from 1 kernel to something like 8.

Comment (ailzhang) [reply torch/nn/functional.py]: Yea but our TH implementation currently doesn't support `requires_grad=True` of `target`. 
@colesbury What do you recommend to solve it here? Is porting smooth_l1_loss to Aten preferred over this hacky workaround?

Comment (colesbury) [reply torch/nn/functional.py]: I misunderstood how _pointwise_loss worked. This is fine.

File: torch/nn/functional.py [hunk 2]
```diff
@@ -1789,10 +1790,10 @@ def smooth_l1_loss(input, target, size_average=None, reduce=None, reduction='ele
...
+        lambda a, b: torch.where(torch.abs(a - b) < 1, 0.5 * (torch.abs(a - b)) ** 2, torch.abs(a - b) - 0.5),
```
Comment (fmassa) [inline torch/nn/functional.py]: nit: if you remove this from a lambda, you could store the result of `torch.abs(a - b)`, which would reduce the number of kernel launches by quite a bit
```python
def _smooth_l1_loss(a, b):
    t = torch.abs(a - b)
    return torch.where(t < 1, 0.5 * t ** 2, t - 0.5)
```

**Comment by SsnL** [comment]
can we have a test?

File: test/test_nn.py [hunk 3]
```diff
@@ -5126,6 +5126,21 @@ def test_pointwise_loss_target_grad_none_reduction(self):
...
+            'smooth_l1_loss': lambda x, y, r: F.smooth_l1_loss(x, y, reduction=r),
+        }
+
+        input = torch.ones(2, 1)
+        for name, fn in losses.items():
+            for requires_grad in (True, False):
+                # When target.requires_grad=True, its impl is in Python, while the other is in TH.
+                target = torch.ones(2, 10, requires_grad=requires_grad)
+                l = fn(input, target, 'none')
+                self.assertEqual(l.size(), target.size())
```
Comment (SsnL) [inline test/test_nn.py]: I think one of the issues you fix is about missing backward computation. Could you also add a gradcheck here?

**Review by ssnl — APPROVED** [review]

**Pull request closed without merging**